lib/block_view_lockups.go

```diff
@@ -0,0 +1,1784 @@
+		return 0, 0, nil, errors.Wrap(RuleErrorCoinUnlockOfAmountZero,
+			"_connectCoinUnlock")
+	}
+
+	// Convert the TransactorPublicKey to HODLerPKID
+	transactorPKIDEntry := bav.GetPKIDForPublicKey(txn.PublicKey)
+	if transactorPKIDEntry == nil || transactorPKIDEntry.isDeleted {
+		return 0, 0, nil, errors.Wrapf(RuleErrorCoinUnlockInvalidHODLerPKID,
+			"_connectCoinUnlock")
+	}
+	hodlerPKID := transactorPKIDEntry.PKID
+
+	// Convert the ProfilePublicKey to ProfilePKID.
+	var profilePKID *PKID
+	if txMeta.ProfilePublicKey.IsZeroPublicKey() {
+		profilePKID = ZeroPKID.NewPKID()
+	} else {
+		profilePKIDEntry := bav.GetPKIDForPublicKey(txMeta.ProfilePublicKey.ToBytes())
+		if profilePKIDEntry == nil || profilePKIDEntry.isDeleted {
+			return 0, 0, nil, errors.Wrapf(RuleErrorCoinUnlockInvalidProfilePKID,
+				"_connectCoinUnlock")
+		}
+		profilePKID = profilePKIDEntry.PKID
+	}
+
+	// Retrieve unlockable locked balance entries.
+	unlockableLockedBalanceEntries, err := bav.GetUnlockableLockedBalanceEntries(
+		hodlerPKID, profilePKID, blockTimestamp)
+	if err != nil {
+		return 0, 0, nil, errors.Wrapf(err, "_connectCoinUnlock")
+	}
+	if len(unlockableLockedBalanceEntries) == 0 {
+		return 0, 0, nil, errors.Wrapf(RuleErrorCoinUnlockNoUnlockableCoinsFound,
+			"_connectCoinUnlock")
+	}
+
+	// Unlock coins until the amount specified by the transaction is deducted.
+	var newLockedBalanceEntries []*LockedBalanceEntry
+	var prevLockedBalanceEntries []*LockedBalanceEntry
+	remainingUnlockBalance := txMeta.CoinsToUnlockBaseUnits.Clone()
+	for _, unlockableLockedBalanceEntry := range unlockableLockedBalanceEntries {
+		newLockedBalanceEntry := unlockableLockedBalanceEntry.Copy()
+		if newLockedBalanceEntry.BalanceBaseUnits.Gt(remainingUnlockBalance) ||
+			newLockedBalanceEntry.BalanceBaseUnits.Eq(remainingUnlockBalance) {
+			remainingUnlockBalance = uint256.NewInt()
+			newLockedBalanceEntry.BalanceBaseUnits =
+				*uint256.NewInt().Sub(&newLockedBalanceEntry.BalanceBaseUnits, remainingUnlockBalance)
+		} else {
+			remainingUnlockBalance =
+				uint256.NewInt().Sub(remainingUnlockBalance, &newLockedBalanceEntry.BalanceBaseUnits)
+			newLockedBalanceEntry.BalanceBaseUnits = *uint256.NewInt()
+		}
+
+		// Append the new LockedBalanceEntry and prev in the event we rollback the transaction.
+		newLockedBalanceEntries = append(newLockedBalanceEntries, newLockedBalanceEntry)
+		prevLockedBalanceEntries = append(prevLockedBalanceEntries, unlockableLockedBalanceEntry)
+
+		// Break if we've satisfied the unlock amount.
+		if remainingUnlockBalance.IsZero() {
+			break
+		}
+	}
+	if !remainingUnlockBalance.IsZero() {
+		return 0, 0, nil, errors.Wrapf(RuleErrorCoinUnlockInsufficientUnlockableCoins,
+			"_connectCoinUnlock")
+	}
+
+	// Update the LockedBalanceEntries.
+	for _, lockedBalanceEntry := range newLockedBalanceEntries {
+		bav._setLockedBalanceEntry(lockedBalanceEntry)
```

Comment:
I think this is a bug in that it will set a bunch of LockedBalanceEntrys in the db with zero balance. Instead, you may want to _delete newBalanceEntrys if they're zero otherwise set them.